File: Algorithms/src/2064. Minimized Maximum of Products Distributed to Any Store/Code/solution.java
class Solution {
    private boolean canDistribute(int[] quantities, int maxProducts, int n) {
        int storesNeeded = 0;
        for (int quantity : quantities) {
            storesNeeded += (int) Math.ceil((double) quantity / maxProducts);
            if (storesNeeded > n)
                return false;
        }
        return storesNeeded <= n;
    }

    public int minimizedMaximum(int n, int[] quantities) {
        int low = 1;
        int high = Arrays.stream(quantities).max().getAsInt();
        int answer = high;

        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (canDistribute(quantities, mid, n)) {
                answer = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return answer;
    }
}

File: Algorithms/src/2064. Minimized Maximum of Products Distributed to Any Store/Code/solution.py
import math
from typing import List

class Solution:
    def minimizedMaximum(self, n: int, quantities: List[int]) -> int:
        def canDistribute(maxProducts):
            storesNeeded = 0
            for quantity in quantities:
                storesNeeded += math.ceil(quantity / maxProducts)
                if storesNeeded > n:
                    return False
            return storesNeeded <= n

        low, high = 1, max(quantities)
        answer = high

        while low <= high:
            mid = (low + high) // 2
            if canDistribute(mid):
                answer = mid
                high = mid - 1
            else:
                low = mid + 1

        return answer